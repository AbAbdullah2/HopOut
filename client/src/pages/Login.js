import React, { useEffect, useState } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { solid } from '@fortawesome/fontawesome-svg-core/import.macro'
import { useNavigate } from 'react-router-dom';
import Header from '../components/Header';
import toast, { Toaster } from 'react-hot-toast';
import { postLogin, getUser } from '../services/api';

function Login(props) {
  const { curUser, setCurUser } = props
  const navigate = useNavigate();

  useEffect(() => {
    if (curUser) navigate('/events');
  });

  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");

  const login = (e) => {
    e.preventDefault();
    const loginData = {
      email: email,
      password: password
    }

    const loginResponse = postLogin(loginData).then(data => {
        if (data.status === 200) {
          // Fetch user details based on userID
            getUser(data.data.data._id).then(userData => {
                setCurUser(userData.data.data);
                navigate('/events');
            })
        }
    }).catch(err => {
      const error = 'Could not login user ' + loginData.email;
      toast.error(error);
      console.log(err)});

    if (!loginResponse) {
      const error = 'Could not login user ' + loginData.email;
      toast.error(error);
    }
  }

  return (
    <div className='bg-stone-100 min-h-screen'>
      <Toaster />
      <div className='mx-auto flex flex-col h-full'>
        <Header icons={false} />
        <div className="flex items-center justify-center py-6">
          <div className="w-full max-w-md">
            <h2 className="mt-6 text-center text-3xl font-bold tracking-tight text-gray-900">
              Welcome back!
            </h2>
            <form className="mt-8 space-y-6" onSubmit={login}>
              <div className="-space-y-px rounded-md shadow-sm">
                <div>
                    <label htmlFor="email" className="sr-only">
                    Email address
                    </label>
                    <input
                    id="email"
                    name="email"
                    type="email"
                    autoComplete="email"
                    required
                    className="relative block w-full appearance-none rounded-none rounded-t-md border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-indigo-500 focus:outline-none focus:ring-indigo-500"
                    placeholder="Email"
                    value={email}
                    onChange={(e) => setEmail(e.target.value)}
                    />
                </div>
                <div>
                    <label htmlFor="password" className="sr-only">
                    Password
                    </label>
                    <input
                    id="password"
                    name="password"
                    type="password"
                    autoComplete="current-password"
                    required
                    className="relative block w-full appearance-none rounded-none rounded-b-md border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-indigo-500 focus:outline-none focus:ring-indigo-500"
                    placeholder="Password"
                    value={password}
                    onChange={(e) => setPassword(e.target.value)}
                    />
                </div>
              </div>
              <div className="flex items-center space-x-5">
                <button
                    type="submit"
                    className="group relative flex w-full justify-center rounded-md border border-transparent bg-blue-400 py-2 px-4 text-sm font-medium text-white hover:bg-slate-800 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2"
                >
                    <span className="absolute inset-y-0 left-0 flex items-center pl-3">
                    <FontAwesomeIcon icon={solid('lock')} className="h-5 w-5 text-white" aria-hidden="true" />
                    </span>
                    Login
                </button>
                <button
                  className="group relative flex w-full justify-center rounded-md border border-transparent bg-blue-400 py-2 px-4 text-sm font-medium text-white hover:bg-slate-800 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2"
                  onClick={(e) => navigate('/signup')}
                >
                  Sign Up 
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
}

export default Login;