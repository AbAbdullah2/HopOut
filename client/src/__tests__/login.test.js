const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

import '@testing-library/jest-dom'
import {render, screen, cleanup, fireEvent} from '@testing-library/react'
import renderer from 'react-test-renderer';
import Login from '../pages/Login'


test('renders login fields', () => {
  render(<Login />);

  const emailInput = screen.getByLabelText('Email address');
  const passwordInput = screen.getByLabelText('Password');
  const loginButton = screen.getByRole('button', { name: 'Login' });
  const signupButton = screen.getByRole('button', { name: 'Sign Up' });

  expect(emailInput).toBeInTheDocument();
  expect(passwordInput).toBeInTheDocument();
  expect(loginButton).toBeInTheDocument();
  expect(signupButton).toBeInTheDocument();
});

test('Clickable buttons', () => {    
  render(<Login />);

  const emailInput = screen.getByLabelText('Email address');
  const passwordInput = screen.getByLabelText('Password');
  const loginButton = screen.getByRole('button', { name: 'Login' });
  const signupButton = screen.getByRole('button', { name: 'Sign Up' });

  fireEvent.change(emailInput, { target: { value: '' } });
  fireEvent.change(passwordInput, { target: { value: '' } });
  fireEvent.click(loginButton);
  fireEvent.click(signupButton);
});


test('matches snapshot', () => {
  const data = {email: "jhop@jhu.edu", password: "password!"};
  const tree = renderer.create(<Login data={data}/>).toJSON();
  expect(tree).toMatchSnapshot();
})