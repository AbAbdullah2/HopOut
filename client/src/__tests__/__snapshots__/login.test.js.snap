// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="bg-stone-100 min-h-screen"
>
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "bottom": 16,
        "left": 16,
        "pointerEvents": "none",
        "position": "fixed",
        "right": 16,
        "top": 16,
        "zIndex": 9999,
      }
    }
  />
  <div
    className="mx-auto flex flex-col h-full"
  >
    <div
      className="w-full h-20 bg-slate-800 p-5 text-white flex flex-row flex-nowrap items-center justify-center text-center items-center"
    >
      <div
        className="flex justify-start content-start items-start left-0"
      >
        <a
          className="mx-2"
          href="/"
        >
          <span
            className="text-4xl font-extrabold italic text-blue-400 mr-1"
          >
            H
          </span>
          <span
            className="text-4xl"
          >
            opOut
          </span>
        </a>
      </div>
    </div>
    <div
      className="flex items-center justify-center py-6"
    >
      <div
        className="w-full max-w-md"
      >
        <h2
          className="mt-6 text-center text-3xl font-bold tracking-tight text-gray-900"
        >
          Welcome back!
        </h2>
        <form
          className="mt-8 space-y-6"
          onSubmit={[Function]}
        >
          <div
            className="-space-y-px rounded-md shadow-sm"
          >
            <div>
              <label
                className="sr-only"
                htmlFor="email"
              >
                Email address
              </label>
              <input
                autoComplete="email"
                className="relative block w-full appearance-none rounded-none rounded-t-md border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-indigo-500 focus:outline-none focus:ring-indigo-500"
                id="email"
                name="email"
                onChange={[Function]}
                placeholder="Email"
                required={true}
                type="email"
                value=""
              />
            </div>
            <div>
              <label
                className="sr-only"
                htmlFor="password"
              >
                Password
              </label>
              <input
                autoComplete="current-password"
                className="relative block w-full appearance-none rounded-none rounded-b-md border border-gray-300 px-3 py-2 text-gray-900 placeholder-gray-500 focus:z-10 focus:border-indigo-500 focus:outline-none focus:ring-indigo-500"
                id="password"
                name="password"
                onChange={[Function]}
                placeholder="Password"
                required={true}
                type="password"
                value=""
              />
            </div>
          </div>
          <div
            className="flex items-center space-x-5"
          >
            <button
              className="group relative flex w-full justify-center rounded-md border border-transparent bg-blue-400 py-2 px-4 text-sm font-medium text-white hover:bg-slate-800 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2"
              type="submit"
            >
              <span
                className="absolute inset-y-0 left-0 flex items-center pl-3"
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-lock h-5 w-5 text-white"
                  data-icon="lock"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  style={Object {}}
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M144 144v48H304V144c0-44.2-35.8-80-80-80s-80 35.8-80 80zM80 192V144C80 64.5 144.5 0 224 0s144 64.5 144 144v48h16c35.3 0 64 28.7 64 64V448c0 35.3-28.7 64-64 64H64c-35.3 0-64-28.7-64-64V256c0-35.3 28.7-64 64-64H80z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </span>
              Login
            </button>
            <button
              className="group relative flex w-full justify-center rounded-md border border-transparent bg-blue-400 py-2 px-4 text-sm font-medium text-white hover:bg-slate-800 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:ring-offset-2"
              onClick={[Function]}
              type="button"
            >
              Sign Up
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
